server/src/test/java/org/elasticsearch/common/settings/SettingTests.java

```diff
@@ -802,6 +803,25 @@ public void testDynamicKeySetting() {
         }
     }
 
+    public void testPrefixKeySettingFallbackAsMap() {
+        Setting.AffixSetting<Boolean> setting = Setting.prefixKeySetting(
+            "foo.",
+            "bar.",
+            (ns, key) -> Setting.boolSetting(key, false, Property.NodeScope)
+        );
+
+        assertTrue(setting.match("foo.bar"));
+        assertTrue(setting.match("bar.bar"));
+
+        Map<String, Boolean> map = setting.getAsMap(Settings.builder().put("foo.bar", "true").build());
+        assertEquals(1, map.size());
+        assertTrue(map.get("bar"));
+
+        map = setting.getAsMap(Settings.builder().put("bar.bar", "true").build());
+        assertEquals(1, map.size());
+        assertTrue(map.get("bar"));
```

Comment:
>>we should prefer the new prefix over the old prefix in getAsMap.
>
> That was a bit much for the implementation in getAsMap, duplicate settings are applied in any order already so I removed that part of the test.

```java
map = setting.getAsMap(Settings.builder().put("foo.bar", "false").put("bar.bar", "true").build());
assertEquals(1, map.size());
assertFalse(map.get("bar"));
```

I'm a bit worried about this not working, you might end up using a fallback rather than the correct setting value without noticing. Wouldn't something like below work, I haven't tested it. But the idea is, if the map already contains a value for namespace, only ever overwrite it the key is prefixed with the default prefix and not the fallback one.

```java
public Map<String, T> getAsMap(Settings settings) {
    Map<String, T> map = new HashMap<>();
    matchStream(settings).distinct().forEach(key -> {
        String namespace = this.key.getNamespace(key);
        if (map.containsKey(namespace) == false || key.startsWith(this.key.prefix)) {
            Setting<T> concreteSetting = getConcreteSetting(namespace, key);
            map.put(namespace, concreteSetting.get(settings));
        }
    });
    return Collections.unmodifiableMap(map);
}
```

Reply:
I was under the mistaken impression that in normal settings the primary was not preferred, so it didn't make sense to add that behavior here.  I confirmed in the case of a primary setting and it's fallback, the primary is preferred.  Added a test to that effect.

I've also made AffixSettings match that behavior and restored the quoted test case.